src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -833,6 +833,71 @@ class Program
             Assert.True(comp.Assembly.RuntimeSupportsByRefFields);
         }
 
+        /// <summary>
+        /// Ref fields of ref struct type are not supported.
+        /// </summary>
+        [Fact]
+        public void RefFieldTypeRefStruct_01()
+        {
+            var source =
+@"#pragma warning disable 169
+ref struct R1<T>
+{
+}
+ref struct R2<T>
+{
+    public ref R1<T> F;
+}
+class Program
+{
+    static void F(ref R1<int> r1)
+    {
+        var r2 = new R2<int>();
+        r2.F = ref r1;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (7,12): error CS9050: A ref field cannot refer to a ref struct.
+                //     public ref R1<T> F;
+                Diagnostic(ErrorCode.ERR_RefFieldCannotReferToRefStruct, "ref R1<T>").WithLocation(7, 12));
+        }
+
+        /// <summary>
+        /// Ref fields of ref struct type are not supported.
+        /// </summary>
+        [Fact]
+        public void RefFieldTypeRefStruct_02()
+        {
+            var sourceA =
+@".class public sealed R1 extends [mscorlib]System.ValueType
+{
+  .custom instance void [mscorlib]System.Runtime.CompilerServices.IsByRefLikeAttribute::.ctor() = (01 00 00 00)
+}
+.class public sealed R2 extends [mscorlib]System.ValueType
+{
+  .custom instance void [mscorlib]System.Runtime.CompilerServices.IsByRefLikeAttribute::.ctor() = (01 00 00 00)
+  .field public valuetype R1& F
```

Comment:
unrelated: do we give a use-site error when a non-`ref struct` contains a ref field?

Reply:
> do we give a use-site error when a non-`ref struct` contains a ref field?

No, we're not reporting a use-site error for use of such a field currently. There is an item in the [test plan](https://github.com/dotnet/roslyn/issues/59194), and I'll add this in a follow-up PR.